python/ray/train/ml_trainer.py

```diff
@@ -0,0 +1,108 @@
+import abc
+import logging
+from typing import Dict, Union, Callable, Optional
+
+from ray.data import Dataset
+from ray.ml.preprocessor import Preprocessor
+from ray.ml.checkpoint import Checkpoint
+from ray.ml.config import ScalingConfig, RunConfig
+from ray.ml.result import Result
+from ray import tune
+from ray.tune.trainable import ConvertibleToTrainable
+from ray.tune.utils import deep_update
+
+GenDataset = Union[Dataset, Callable[[], Dataset]]
+
+
+logger = logging.getLogger(__name__)
+
+
+# TODO: support conditional hyperparameter tuning of Trainer __init__ args.
```

Comment:
Move this TODO to a more appropriate location instead of at the top of the class.

(also, doesn't _override_attributes_with_config do this?)

Reply:
I don't think so, please see the discussion on Slack. Conditional hyperparameter tuning is not supported I believe.

Reply:
> Have been chatting with @Yard1 on this for a bit, but it seems like with the current API it’s not possible to tune Trainer args in conjunction with each other?
For example, I want to do a grid search over 3 different datasets, but for each dataset I have a specific preprocessor. This doesn’t seem possible unless we combine the dataset and preprocessor into a single argument.
This is because
With the new API, we are now enforcing that the keys for the tune config dict have to match with the arg names for Trainer init, and the values have to match with what Trainer init expects.
Normally, the recommendation for joint hyperparameters is to combine them into one hyperparameter and use a tuple, but this is no longer possible with constraint in 1